File: local_development/knn.r
#! /usr/bin/env Rscript
#
#
#
#
#  some code sourced from: http://stackoverflow.com/questions/14025176/finding-k-nearest-neighbor-for-only-one-point-not-an-entire-matrix-using-r
#
#
# Note: the FNN package requires all data be loaded into memory before it can run.
#       with web scale data, this is bad. see the included ./knn.py file
#       for an example of constant RAM stream processing. 
#
#       to scale this implementation as production code, we'd need to
#       split lines from standard in into batches, where each batch fits
#       in memory. the code is somewhat awkward and beyond the scope of 
#       this example.
#
#       the code that blocks streaming can be found in the FNN source code in
#       KNN_ANN.cpp lines 188-190
#       
#        	  //copy data into  matrix from a vector column by column
#	          Rvector2ANNarray(data_pts, data, n, d);
# 	        Rvector2ANNarray(query_pts, query, m, d);
#
#

require(FNN)

# config values
train_fpath      <- './data/train.tsv'
test_fpath       <- './data/test.tsv'
k                <- 3

# load training data into memory
df.train         <- read.table(train_fpath, header=F, sep='\t')

# load test data into memory
df.test          <- read.table('stdin',  header=F, sep='\t')

# separate keys and values
df.train.keys    <- df.train[,1]
df.test.keys     <- df.test[,1]

df.train.values  <- df.train[,-1]
df.test.values   <- df.test[,-1]

# run KNN
results <- knnx.index(df.train.values, df.test.values, k=k)

# push the KNN assignment results to standard out
for(i in 1:dim(results)[1]) {
  key     <- df.test.keys[i]
  nearest <- results[i,]

  cat(paste(c('R KNN:',key, '->', nearest,'\n')))
  }

File: local_development/knn.py
#!/usr/bin/python
'''
    some code sourced from: http://stackoverflow.com/questions/12369484/searching-for-k-nearest-points
'''

import numpy
from sklearn.neighbors import NearestNeighbors
import string
import sys


class KeyedSampleRow:
  ''' encapsulates a row of randomly generated data '''
  def __init__(self, line=None):
    '''transforms a line of text into a (key, numpy.array) tuple.
       assumes the first column of a tab delimited file is an integer key
       and remaining columns are float values '''
    row        = map(string.strip, line.split('\t'))
    self.key   = int(row[0])
    self.value = numpy.array(map(float, row[1:]))
    pass
  def __str__(self):
     return  
  pass


def main ():
  ''' builds a KNN model from training data, then reads from standard in and assigns neighbors to all incoming sample rows  '''

  # config values
  train_fpath = './data/train.tsv'
  k = 3
  
  # load training data into memory, use it to build a KNN model
  training_data = [KeyedSampleRow(l) for l in open(train_fpath, 'r')]

  Train = numpy.matrix(map(lambda t: t.value, training_data))
  knn = NearestNeighbors(n_neighbors=k)
  knn.fit(Train)
  
  # stream read Test data from standard in
  # Note: this is a iterator of (key, numpy.array) tuples.
  #       it uses a small, test data scale independent amount of memory
  Test = (KeyedSampleRow(l) for l in sys.stdin)

  # online-query the KNN model
  for test_row in Test:
    nearest = knn.kneighbors(test_row.value, return_distance=False)
    
    # note: to match R indexing, +1 results
    print 'py KNN:\t{0} -> {1}'.format(test_row.key, nearest+1)

  print
  pass

if __name__ == '__main__':
  main()
